specify/specify7
# Allow catalogNumber format to be determined by CollectionObjectType

Fixes #5232, #5233 
See #5234

⚠️ **Note:** This PR affects database migrations. See [migration testing instructions](https://specify.github.io/testing/pull_requests.html#prs-tagged-with-label-migration).


Aside from implementing #5232 and #5233, there are some other improvements in this PR!

**Schema Config**

- When selecting a Field Format in the Schema Configuration tool, Specify now groups field formatters for the field they selected (Field Formats that explicitly reference the field in `fieldName=fieldname`) above the rest

<img width="845" alt="Screenshot 2024-12-03 at 3 08 19 PM" src="https://github.com/user-attachments/assets/559439b1-e7a9-436f-830a-45776586cc9b">

**Migrations**
There were some bugs in other migrations which used the functions defined in [specifyweb/specify/update_schema_config.py](https://github.com/specify/specify7/blob/production/specifyweb/specify/update_schema_config.py). These bugs have been fixed in this PR, and a migration has been written to fix these bugs for databases already affected: [specify/0017_schemaconfig_fixes.py](https://github.com/specify/specify7/blob/issue-5232/specifyweb/specify/migrations/0017_schemaconfig_fixes.py)

**UIFormatter picklists for Collection -> catalogNumFormatName and CollectionObjectType -> catalogNumberFormatName**
- Everywhere in the application*, these fields will now appear as a picklist whose items are the names of `CO -> catalogNumber` Field Formatters

<img width="1287" alt="Screenshot 2024-12-03 at 3 21 02 PM" src="https://github.com/user-attachments/assets/0ea9f366-566d-49b6-95dd-cb100fec7116">

*on the DataEntry Form, they will only appear as a PickList if `uitype="combobox"` is used on the form definition

### Checklist

      and self-explanatory (or properly documented)

### Testing instructions

- Define a Field Format for `Collection Object -> catalogNumber` which is different than the current `Collection -> Catalog Num Format Name`
  - An example field format called `Rocks` is provided below:  
```xml
<format system="false" name="Rocks" class="edu.ku.brc.specify.datamodel.CollectionObject" fieldname="catalogNumber" default="false">
	<autonumber>edu.ku.brc.af.core.db.AutoNumberGeneric</autonumber>
	<field type="constant" size="4" value="ROCK"/>
	<field type="separator" size="1" value="-"/>
	<field type="numeric" size="6" inc="true"/>
</format>
```
- Go to the Data Entry form to create a new Collection Object Type
- [ ] Ensure the Catalog Number Format Name picklist shows all Catalog Number Field Formats 
- Save the Collection Object Type with a Catalog Number Format Name that is different than the Collection -> catalogNumFormatName
- Make sure `collectionObjectType` is on the Collection Object form 
- Go to the Data Entry form to create a new Collection Object 
- Change the Collection Object Type of the Collection Object to be that which you previously saved
- [ ] Ensure the Catalog Number of the Collection Object is replaced with the pattern of the new 
- (If needed) Modify the Catalog Number so that it matches the format of formatter
- [ ] Save the Collection Object with the new format
